### DISCUSSION TITLE
how swap from 1 token to usdt without direct pair? and fees on swap from token to eth

### DISCUSSION THREAD

**Discussion by facilficom** [Elnaril/uniswap-universal-router-decoder#67] [Q&A] [answered]
question 1:
I am trying the possibility of performing a swap from [Token] to USDT. However, I have noticed that there isn't a direct trading pair available for this token and USDT. The swap with 3 paths is not working for me. 

question 2:
Is it possible to swap the fees to WETH and then unwrap WETH to ETH in the same transaction? If so, how can I achieve this using the Universal Router? i can take the fee from amountout like explain in a discussion before this but i cannot swap to weth and then eth in the same transaction

**Comment by Elnaril** [comment]
Hi @facilficom !
For both questions, could you share the code you tried so far?

**Reply by facilficom** [reply]
here, is for other conditions: 
    path = [token_sold, token_bought]
    recipient_rate = int((1 - COMMISSION_RATE) * 10000)
    if eth_value:
        chain_builder = chain_builder.v2_swap_exact_in(FunctionRecipient.ROUTER, amount_sold, 0, path, payer_is_sender=True)
        chain_builder = chain_builder.unwrap_weth(FunctionRecipient.ROUTER, 0)
        chain_builder = chain_builder.pay_portion(FunctionRecipient.SENDER, '0x0000000000000000000000000000000000000000', recipient_rate)
        chain_builder = chain_builder.sweep(FunctionRecipient.CUSTOM, '0x0000000000000000000000000000000000000000', 0, OWNER_WALLET_ADDRESS)
    else:  
        chain_builder = chain_builder.v2_swap_exact_in(FunctionRecipient.ROUTER, amount_sold, 0, path, payer_is_sender=True)
        chain_builder = chain_builder.pay_portion(FunctionRecipient.SENDER, token_bought, recipient_rate)
        if token_bought.lower() == weth_address.lower():
            chain_builder = chain_builder.unwrap_weth(FunctionRecipient.CUSTOM, 0, OWNER_WALLET_ADDRESS)
        else:
            chain_builder = chain_builder.sweep(FunctionRecipient.CUSTOM, token_bought, 0, OWNER_WALLET_ADDRESS)

    encoded_input = chain_builder.build(codec.get_default_deadline())

    trx_params = {
        "from": account_checksum,
        "to": uniswap_address,
        "maxPriorityFeePerGas": maxpriorityfeepergas,
        "maxFeePerGas": maxfeepergas,
        "type": '0x2',
        "chainId": chain_id,
        "value": 0,
        "nonce": nonce,
        "data": encoded_input,
    }
    trx_params_gas_estimate = web3.eth.estimate_gas(trx_params)
    trx_params.update({'gas': int(trx_params_gas_estimate * 1.2)})
    signed_transaction = web3.eth.account.sign_transaction(trx_params, private_key)
    trx_hash = web3.eth.send_raw_transaction(signed_transaction.raw_transaction)
    print(f"Transaction Hash: {trx_hash.hex()}")
    receipt = web3.eth.wait_for_transaction_receipt(trx_hash)
    print(f"Transaction receipt: {receipt}")
after this:
        else:
            chain_builder = chain_builder.sweep(FunctionRecipient.CUSTOM, token_bought, 0, OWNER_WALLET_ADDRESS)
will be useful to swap the remain for eth becuase if its not possible, the owner wallet will have thousands of tokens.

in the while, if i make a path of 3 tokens (usdt, wbnb, token or viceversa) the transaction doesnt work

**Comment by Elnaril** [comment]
Ok, thanks!

So, if I understand correctly what you try to achieve, you'd like to swap the remaining tokens to `WETH` and unwrap to `ETH` instead of `sweep`ing them in the following code block, right ?
```python
else:
    chain_builder = chain_builder.sweep(FunctionRecipient.CUSTOM, token_bought, 0, OWNER_WALLET_ADDRESS)
```
In this case, if you have the right path, you can use the method `v2_swap_exact_in_from_balance()` to encode the swap to `WETH` using the full amount in the router's balance, then `unwrap_weth()`.

The tricky part would be to have the right path. For that, you may want to give a try to the: [uniswap-smart-path](https://github.com/Elnaril/uniswap-smart-path)

--- 

For your issue with a 3-token path, could you please share a minimal example and a link to the failed transaction?

**Reply by facilficom** [reply]
yes, im trying but if i add swap in from balance the transaction give me an error, i dont think its possible to make 2 swap in the same data.  for the 3 token path example, it seems on a universal router they passed the first amountout swap to the address pair and then make the other swap, i think this is the problem. https://bscscan.com/tx/0xff8b9f7daab571a44b7ba952c6b7fb471237f58975d0b16ac06a283f244e9e48#eventlog

**Reply by Elnaril** [reply, marked as answer]
> for the 3 token path example, it seems on a universal router they passed the first amountout swap to the address pair and then make the other swap, i think this is the problem. https://bscscan.com/tx/0xff8b9f7daab571a44b7ba952c6b7fb471237f58975d0b16ac06a283f244e9e48#eventlog

This is the decoded input data used in the transaction you linked:

```python
(<Function execute(bytes,bytes[],uint256)>,
 {'commands': b'\x08',
  'inputs': [(<Function V2_SWAP_EXACT_IN(address,uint256,uint256,address[],bool)>,
              {'recipient': '0xEd4FF665F61B36A63dFA260F6C8CcFB389d79C19',
               'amountIn': 162262162497017081,
               'amountOutMin': 386689736694041613984,
               'path': ['0x55d398326f99059fF775485246999027B3197955',
                        '0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c',
                        '0xF3c7CECF8cBC3066F9a87b310cEBE198d00479aC'],
               'payerIsSender': True},
              {'revert_on_fail': True})],
  'deadline': 1721930957})

```
As you can see, there is a 3-token path. There is 2 swap events in the transaction logs because there is one per pool.


> i dont think its possible to make 2 swap in the same data

If you look into the [integration tests](https://github.com/Elnaril/uniswap-universal-router-decoder/tree/master/integration_tests), you'll find some of them doing 2 swaps in the same transaction.
